test_runner/src/main/kotlin/ftl/gc/GcStorage.kt

```diff
@@ -99,43 +103,44 @@ object GcStorage {
 
     private fun upload(file: String, fileBytes: ByteArray, rootGcsBucket: String, runGcsPath: String): String {
         val fileName = Paths.get(file).fileName.toString()
-        val gcsFilePath = GCS_PREFIX + join(rootGcsBucket, runGcsPath, fileName)
-
-        // 404 Not Found error when rootGcsBucket does not exist
-        val fileBlob = BlobInfo.newBuilder(rootGcsBucket, join(runGcsPath, fileName)).build()
-
-        val progress = ProgressBar()
-        try {
-            progress.start("Uploading $fileName")
-            storage.create(fileBlob, fileBytes)
-        } catch (e: Exception) {
-            fatalError(e)
-        } finally {
-            progress.stop()
+        return uploadCache[fileName] ?: uploadCache.computeIfAbsent(fileName) {
+            val gcsFilePath = GCS_PREFIX + join(rootGcsBucket, runGcsPath, fileName)
+
+            // 404 Not Found error when rootGcsBucket does not exist
+            val fileBlob = BlobInfo.newBuilder(rootGcsBucket, join(runGcsPath, fileName)).build()
+
+            val progress = ProgressBar()
+            try {
+                progress.start("Uploading $fileName")
+                storage.create(fileBlob, fileBytes)
+            } catch (e: Exception) {
+                fatalError(e)
+            } finally {
+                progress.stop()
+            }
+            gcsFilePath
         }
-
-        return gcsFilePath
     }
 
     fun download(gcsUriString: String, ignoreError: Boolean = false): String {
         val gcsURI = URI.create(gcsUriString)
         val bucket = gcsURI.authority
         val path = gcsURI.path.drop(1) // Drop leading slash
-
-        val outputFile = File.createTempFile("tmp", null)
-        outputFile.deleteOnExit()
-
-        try {
-            val blob = storage.get(bucket, path)
-            val readChannel = blob.reader()
-            val output = FileOutputStream(outputFile)
-            output.channel.transferFrom(readChannel, 0, Long.MAX_VALUE)
-            output.close()
-        } catch (e: Exception) {
-            if (ignoreError) return ""
-            fatalError(e)
+        return downloadCache[path] ?: downloadCache.computeIfAbsent(path) {
+            val outputFile = File.createTempFile("tmp", null)
+            outputFile.deleteOnExit()
+
+            try {
+                val blob = storage.get(bucket, path)
+                val readChannel = blob.reader()
+                val output = FileOutputStream(outputFile)
```

Comment:
I think we could take advantage of `use()` extension function
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html
This ensures closing stream in case when exception was ignored (and prevent memory leak)
Let me know what you think  :)